### ISSUE TITLE
Views created in no_grad block still have requires_grad=True

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#11390]
```
>>> x = torch.randn(3, requires_grad=True)
>>> with torch.no_grad():
...   print(x[1])
...
tensor(-0.0967, requires_grad=True)
```

**Comment by vishwakftw** [comment]
Is the example in `grad_mode.py` suggesting that this behavior is correct?

**Comment by t-vi** [comment]
It's the output, but it's unexpected.

**Comment by gchanan** [comment]
@SsnL are you still looking at this?

**Comment by SsnL** [comment]
@gchanan No, we never reached an agreement on what to do with it.  cc people who were involved in the discussion @apaszke @colesbury @soumith .

**Comment by apaszke** [comment]
IIRC we did arrive at the conclusion that it was the desired behavior?

**Comment by apaszke** [comment]
Oh, we only wanted to make `.requires_grad` to always return `False` and remove that from printing inside the `no_grad` block to avoid confusion. But that was it if I remember correctly.

**Comment by gchanan** [comment]
was that discussion captured anywhere?

**Comment by SsnL** [comment]
@gchanan Partly in https://github.com/pytorch/pytorch/pull/12502, and partly on slack in a group chat. IIRC, the core of the discussion is about this pattern:
```py
z = torch.ones(..., requires_grad=True)
with torch.no_grad():
  x = z[0]
x[i] = y
```

and whether you should be able to backprop from `z` to `y`.

**Comment by gchanan** [comment]
@apaszke's final reply in https://github.com/pytorch/pytorch/pull/12502 seems different than the conclusion here:

> @ezyang I do think that we should change the semantics of detach() to match what I said closer, and then simply say "every tensor allocated in a no grad block is as if detach() was called on it". This would give you the semantics I described.
> This is what we've converged on in today's meeting with @SsnL and @soumith.

i.e. the quote suggests changes to `detach`, but the suggestion here:

> Oh, we only wanted to make .requires_grad to always return False and remove that from printing inside the no_grad block to avoid confusion. But that was it if I remember correctly.

doesn't have any changes to `detach`.  Is one more updated that the other or are these two separate issues?

**Comment by ezyang** [comment]
> Oh, we only wanted to make .requires_grad to always return False and remove that from printing inside the no_grad block to avoid confusion. But that was it if I remember correctly.

I think the final reply in #12502 is more in line with what I would have expected.

**Comment by ezyang** [comment]
In the mean time since this issue was posted, Alban has added an error message forbidding this case:

```
>>> z = torch.ones(2, 2)
>>> with torch.no_grad():
...   x = z[0]
... 
>>> y = torch.ones((), requires_grad=True)
>>> x[0] = y
__main__:1: UserWarning: A view was created in no_grad mode and is being modified inplace with grad mode enabled. Given that this use case is ambiguous and error-prone, it is deprecated and will be forbidden  starting 1.6 (see https://github.com/pytorch/pytorch/pull/32839 for more details about this). You can clarify your code and remove this warning by moving both the view and the inplace either both inside the no_grad block (if you don't want the inplace to be tracked) or both outside (if you want the inplace to be tracked). (Triggered internally at  ../torch/csrc/autograd/variable.cpp:547.)
```

**Comment by albanD** [comment]
Note that this warning will become an error very soon we're aiming for 1.9

The question still remains open if `x.requires_grad` should be `True` or not?

**Comment by ezyang** [comment]
The view still requires grad:

```
>>> import torch
>>> z = torch.ones(2, 2, requires_grad=True)
>>> with torch.no_grad():
...   x = z[0]
... 
>>> x
tensor([1., 1.], requires_grad=True)
```

**Comment by hjmshi** [comment]
Hi, have there been any follow-ups to this issue? This is blocking our usage of `torch.compile` from PyTorch 2 in our Distributed Shampoo implementation. Thanks in advance!

See: https://github.com/pytorch/pytorch/issues/99968#issuecomment-1809347254.

cc: @tsunghsienlee @shintaro-iwasaki @minddrummer @csmiler @mlazos @bdhirsh @yuchenhao

**Comment by albanD** [comment]
Hi,

There isn't much to do here tbh. This behavior is pretty much expected. In spirit, the requires-grad-ness is a property of entry in memory, not the Tensor. So if any Tensor viewing into a slot in memory requires grad, then that slot requires grad and any other Tensor looking into that memory will also require gradients. Whether or not you created it in no-grad mode or not.

The error in the other comment is also expected: the behavior when mixing these two modes is unclear as it not clear if the view should be differentiated through or not. So we error out.
Also if you error only happens when using torch.compile but not without, please open a new issues as it is an issue with torch.compile which should not change the behavior of pytorch in any way.

**Comment by ssnl** [comment]
@albanD Maybe not changing behavior, but I think something should be done eventually about this. In PyTorch, a user can generally be agnostic of whether an op outputs a view or not (e.g., reshape, flatten), but should always stay aware of whether something requires_grad. This mindset is what most PyTorch users hold. Therefore, the behavior here can be surprising. 

```py
with torch.no_grad():
   y = x.reshape(3, 2)
```
Having `y.requires_grad` depending on the strides of `x` is pretty unexpected to many.

**Comment by t-vi** [comment]
> In spirit, the requires-grad-ness is a property of entry in memory, not the Tensor. So if any Tensor viewing into a slot in memory requires grad, then that slot requires grad and any other Tensor looking into that memory will also require gradients. 

Could this be a bit more subtle to the casual user when detach give a view that does not require gradients?

**Comment by ezyang** [comment]
For the reshape case, I'm hoping that eventually (maybe in a few years) we can switch reshape to copy-on-write by default, so it always returns "not a view" and then we will consistently return requires_grad=False here.

cc @kurtamohler

**Comment by albanD** [comment]
> Could this be a bit more subtle to the casual user when detach give a view that does not require gradients?

Detach is explicitly asking for this behavior: take a view of that memory that doesn't share any history with what is currently there. It is as-if you created a new slot in memory that you can now start having mutliple views into.
So I don't think it break the mental model even though it does make it significantly more complex.

**Comment by hjmshi** [comment]
Thanks @albanD for the quick response! Yes, this error only occurs when we apply `torch.compile` on our function and does not error out when we run without it. Note that the error only occurs _after_ the parameter has been modified by an in-place operation through its view, but not before. In other words, `torch.compile` and printing will work fine in the first iteration, but fail in the second iteration of the optimizer. I'm happy to create a new issue for this if that would be helpful. 

I also want to emphasize that PyTorch errors out when printing (which should be a read-only function) but not for logging for this same code snippet. Is this really expected? What is the difference between these two cases? Thanks again!

**Comment by albanD** [comment]
> I also want to emphasize that PyTorch errors out when printing (which should be a read-only function) but not for logging for this same code snippet.

depends how you log, but most likely you don't try to access the `grad_fn` when logging which we do during printing. I agree we should fix printing, updated https://github.com/pytorch/pytorch/issues/99968 to that effect.

For `torch.compile` yes please open a separate issue to avoid two discussions in this one.
